pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -858,105 +858,186 @@ func TestPriorityQueue_Pop(t *testing.T) {
-	}
+				return unschedulablePodInfo.Pod, unschedulablePodInfo.Pod
+			},
+			validateFunc: func(q *PriorityQueue, expectedPod *v1.Pod) {
+				if len(q.unschedulablePods.podInfoMap) != 0 {
+					t.Error("Expected unschedulablePods to be empty.")
+				}
+				if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(expectedPod)); !exists {
+					t.Errorf("Expected: %v to be added to activeQ.", expectedPod.Name)
+				}
+			},
+		},
+		{
+			name: "Updating a pod that is in backoff queue and it is still backing off. pod will not be moved to active queue, and it will be updated in backoff queue",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
+				if err := q.podBackoffQ.Add(podInfo); err != nil {
+					t.Errorf("adding pod to backoff queue error: %v", err)
+				}
 
-	// To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before testing AddUnschedulableIfNotPresent.
-	err = q.activeQ.Add(podInfo)
-	if err != nil {
-		t.Fatalf("unexpected error from activeQ.Add: %v", err)
-	}
-	if p, err := q.Pop(); err != nil || p.Pod != medPriorityPodInfo.Pod {
-		t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, p.Pod.Name)
-	}
-	// updating a pod which is in unschedulable queue, and it is not backing off,
-	// we will move it to active queue
-	err = q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(medPriorityPodInfo.Pod, "plugin"), q.SchedulingCycle())
-	if err != nil {
-		t.Fatalf("unexpected error from AddUnschedulableIfNotPresent: %v", err)
-	}
-	if len(q.unschedulablePods.podInfoMap) != 1 {
-		t.Error("Expected unschedulablePods to be 1.")
+				return podInfo.Pod, podInfo.Pod
+			},
+			validateFunc: func(q *PriorityQueue, expectedPod *v1.Pod) {
+				rawPodInfo, err := q.podBackoffQ.Pop()
+				podGotFromBackoffQ := rawPodInfo.(*framework.QueuedPodInfo).Pod
+				if err != nil || podGotFromBackoffQ != expectedPod {
+					t.Errorf("Expected: %v after Pop, but got: %v", expectedPod.Name, podGotFromBackoffQ.Name)
+				}
+			},
+		},
+		{
+			name: "updating a pod which is in unschedulable queue, and it is backing off, it will stay backoff queue",
+			prepareFunc: func(q *PriorityQueue) (oldPod, newPod *v1.Pod) {
+				podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
+				if err := q.activeQ.Add(podInfo); err != nil {
+					t.Fatalf("unexpected error from activeQ.Add: %v", err)
+				}
+				if p, err := q.Pop(); err != nil || p.Pod != medPriorityPodInfo.Pod {
+					t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, p.Pod.Name)
+				}
+				err := q.AddUnschedulableIfNotPresent(logger, q.newQueuedPodInfo(medPriorityPodInfo.Pod, "plugin"), q.SchedulingCycle())
+				if err != nil {
+					t.Fatalf("unexpected error from AddUnschedulableIfNotPresent: %v", err)
+				}
+				if len(q.unschedulablePods.podInfoMap) != 1 {
+					t.Error("Expected unschedulablePods to be 1.")
+				}
```

Comment:
Can we push this Pod into `.unschedulablePods.podInfoMap` directly so that we can simplify prepareFunc? or would it cause some troubles?

Reply:
The ultimate goal of all the above steps is to construct a scenario where the pod is already in unschedulablePods, so we push this Pod into .unschedulablePods.podInfoMap directly can achieve the same effect. But Thare are some differences, as the origian comments syas `To simulate the pod is failed in scheduling in the real world, Pop() the pod from activeQ before testing AddUnschedulableIfNotPresent.`, The original steps are simulations of the scheduling process, which are more in line with the actual scheduling process, but in the `TestPriorityQueue_Update` function we are testing the `update`, so as long as the same effect is achieved, we can do this